# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, we'll gain practice combining DataFrames through concatenation.  We'll also learn to read in tables from a database and store them in separate tables, as well as how to execute various types of joins to selectively combine the information stored in the tables!


## Objectives:
You will be able to:
* Understand and explain when to use DataFrame joins and merges
* Be able to use pd.merge when combining DataFrames based on column values
* Understand, explain and use a range of DataFrame merge types: outer, inner, left and right
* Use pd.concat() to stack DataFrames

## Getting Started

We'll start with a quick section to help us gain practice with concatenating datasets using `pd.concat()`.

### Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [1]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that we have multiple DataFrames to work with, we can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.  

In [3]:
to_concat = [df1, df2, df3]
big_df = pd.concat(to_concat)
big_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="er1.png">

### Setting Join Conditions With Concatenation

We can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, we'll execute an inner join. Before we do, we need to create another table that contains some overlapping index values with a DataFrame that already exists.  

Run the cell below to create the new DataFrame.

In [4]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

Now, in the cell below, use the `pd.concat()` method to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [13]:
to_concat1 = [df1,df4]
df1_and_4 = pd.concat(to_concat1, axis=1, join = 'inner')
df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='er2.png'>

We'll notice that in this case, the results returned contain only the rows with indexes that exist in both tables--rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that we can make full use of the `concat()` functionality in pandas to join DataFrames together--these are just a few of the most common examples pulled from the pandas documentation on the subject.  For a full view of all the ways we can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Loading In Data From SQL Database Tables

Now, we'll move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the  [kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table--but not all--correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  We'll make use of these keys to **_join_** the different tables together into a single DataFrame.  We'll also experiment with different types of joins to help us decide exactly what information we want to combine.  

Before we can begin working with the database, we'll first need to connect to it.  For this, we'll use the `sqlalchemy` library.

In the cell below:

* Import the `create_engine` module from `sqlalchemy`
* Create an engine that connects to `sqlite:///database.sqlite`.

In [16]:
from sqlalchemy import create_engine
engine = create_engine('sqlite:///database.sqlite')
engine

Engine(sqlite:///database.sqlite)

Now that we've successfully connected to our Database, we'll make use of pandas to read in each table individually.  

In the cell below, read in each sql table into separate pandas DataFrames using the `read_sql_table` method.  

The first argument should be a string corresponding to the name of the table, and the second argument should be the `engine` object we created in the cell above. 

If you're unsure of how to do this, see the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_sql_table.html).

In [22]:
cards_df = pd.read_sql_table('cards',engine)
dust_df = pd.read_sql_table('dust_costs',engine)
entourages_df = pd.read_sql_table('entourages',engine)
mechanics_df = pd.read_sql_table('mechanics',engine)
play_requirements_df = pd.read_sql_table('play_requirements',engine)
play_requirements_df

card_id               play_requirement  value
0           KARA_00_07           REQ_NUM_MINION_SLOTS      1
1             PRO_001a           REQ_NUM_MINION_SLOTS      1
2              NAX1_01           REQ_NUM_MINION_SLOTS      1
3          DS1h_292_H1                REQ_STEADY_SHOT      0
4          DS1h_292_H1       REQ_MINION_OR_ENEMY_HERO      0
5              KAR_076             REQ_TARGET_TO_PLAY      0
6               AT_024             REQ_TARGET_TO_PLAY      0
7               AT_024           REQ_TARGET_WITH_RACE     15
8               AT_024              REQ_MINION_TARGET      0
9             BRMA06_2           REQ_NUM_MINION_SLOTS      1
10             NAX5_02      REQ_MINIMUM_ENEMY_MINIONS      1
11             OG_080c            REQ_FRIENDLY_TARGET      0
12             OG_080c             REQ_TARGET_TO_PLAY      0
13             OG_080c              REQ_MINION_TARGET      0
14             CS2_039             REQ_TARGET_TO_PLAY      0
15             CS2_039              REQ_MINION_TARGET      0
16   TB_KaraPortal_002              REQ_MINION_TARGET      0
17             CS2_094             REQ_TARGET_TO_PLAY      0
18     TB_SPT_DPromoHP              REQ_MINION_TARGET      0
19            BRMA15_3           REQ_NUM_MINION_SLOTS      1
20             CS2_141        REQ_TARGET_IF_AVAILABLE      0
21             CS2_141             REQ_NONSELF_TARGET      0
22             CS2_063             REQ_TARGET_TO_PLAY      0
23             CS2_063              REQ_MINION_TARGET      0
24             CS2_063               REQ_ENEMY_TARGET      0
25            GVG_055e            REQ_FRIENDLY_TARGET      0
26              TB_006             REQ_TARGET_TO_PLAY      0
27              TB_006              REQ_MINION_TARGET      0
28          CFM_621t21      REQ_MINIMUM_TOTAL_MINIONS      1
29          LOEA06_04h             REQ_TARGET_TO_PLAY      0
..                 ...                            ...    ...
746            GVG_011              REQ_MINION_TARGET      0
747            CFM_061             REQ_TARGET_TO_PLAY      0
748      AT_132_HUNTER                REQ_STEADY_SHOT      0
749      AT_132_HUNTER       REQ_MINION_OR_ENEMY_HERO      0
750            GVG_047              REQ_MINION_TARGET      0
751            GVG_047               REQ_ENEMY_TARGET      0
752            EX1_278             REQ_TARGET_TO_PLAY      0
753            CS2_012             REQ_TARGET_TO_PLAY      0
754            CS2_012               REQ_ENEMY_TARGET      0
755            EX1_578             REQ_TARGET_TO_PLAY      0
756            EX1_578              REQ_MINION_TARGET      0
757         KARA_00_10  REQ_SECRET_CAP_FOR_NON_SECRET      0
758         KARA_13_11      REQ_MINIMUM_ENEMY_MINIONS      2
759            GVG_120        REQ_TARGET_IF_AVAILABLE      0
760            GVG_120           REQ_TARGET_WITH_RACE     20
761         LOEA02_10a           REQ_NUM_MINION_SLOTS      1
762            CS1_130             REQ_TARGET_TO_PLAY      0
763             AT_025      REQ_MINIMUM_ENEMY_MINIONS      2
764            CS2_061             REQ_TARGET_TO_PLAY      0
765             AT_084            REQ_FRIENDLY_TARGET      0
766             AT_084        REQ_TARGET_IF_AVAILABLE      0
767             AT_084              REQ_MINION_TARGET      0
768           BRMA16_3            REQ_WEAPON_EQUIPPED      0
769           BRMA16_3             REQ_TARGET_TO_PLAY      0
770           BRMA16_3              REQ_MINION_TARGET      0
771            EX1_624             REQ_TARGET_TO_PLAY      0
772             AT_044             REQ_TARGET_TO_PLAY      0
773             AT_044              REQ_MINION_TARGET      0
774      AT_132_SHAMAN           REQ_NUM_MINION_SLOTS      1
775            CS2_031             REQ_TARGET_TO_PLAY      0

[776 rows x 3 columns]

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the dataframes we created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the head of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since we are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [24]:
cards_df_indexed = cards_df.set_index('card_id', True)
cards_df_indexed

player_class         type  \
card_id                                                   
KARA_00_07                            MAGE        SPELL   
NEW1_008a                            DRUID        SPELL   
BRM_010t2                            DRUID       MINION   
AT_132                             NEUTRAL       MINION   
OG_141                             NEUTRAL       MINION   
CRED_17                            NEUTRAL       MINION   
OG_337e                            NEUTRAL  ENCHANTMENT   
BRMA08_1H                          NEUTRAL         HERO   
GVG_011a                           NEUTRAL  ENCHANTMENT   
NAX8_02H_TB                        NEUTRAL   HERO_POWER   
AT_131                             NEUTRAL       MINION   
CS2_122e                           NEUTRAL  ENCHANTMENT   
LOEA05_02ha                        NEUTRAL   HERO_POWER   
CS2_102_H1_AT_132                  WARRIOR   HERO_POWER   
CFM_334e                           NEUTRAL  ENCHANTMENT   
EX1_583                            NEUTRAL       MINION   
TB_PickYourFate_11rand             NEUTRAL        SPELL   
LOE_007t                           WARLOCK        SPELL   
CFM_095                            NEUTRAL       MINION   
BRM_018                            PALADIN       MINION   
PRO_001a                           NEUTRAL        SPELL   
BRMA10_5H                          NEUTRAL       MINION   
BRMA11_1                           NEUTRAL         HERO   
OG_027                              SHAMAN        SPELL   
NAX1_01                            NEUTRAL         HERO   
GVG_030                              DRUID       MINION   
CFM_094                            WARLOCK        SPELL   
DS1h_292_H1                         HUNTER   HERO_POWER   
CS2_227                            NEUTRAL       MINION   
KAR_062                            NEUTRAL       MINION   
...                                    ...          ...   
LOEA12_2                           NEUTRAL   HERO_POWER   
AT_084                             NEUTRAL       MINION   
KARA_00_03                            MAGE         HERO   
BRMA16_3                           NEUTRAL        SPELL   
tt_010a                               MAGE       MINION   
EX1_559                               MAGE       MINION   
EX1_624                             PRIEST        SPELL   
OG_044a                              DRUID       MINION   
AT_044                               DRUID        SPELL   
AT_132_SHAMAN                       SHAMAN   HERO_POWER   
CFM_621t29                         NEUTRAL        SPELL   
EX1_355e                           PALADIN  ENCHANTMENT   
LOEA09_3d                          NEUTRAL   HERO_POWER   
TB_Coopv3_105                      NEUTRAL       MINION   
BRMA13_4                           NEUTRAL   HERO_POWER   
EX1_345t                            PRIEST       MINION   
EX1_009                            NEUTRAL       MINION   
CRED_32                            NEUTRAL       MINION   
FP1_030e                           NEUTRAL  ENCHANTMENT   
TB_PickYourFate_10_Ench            NEUTRAL  ENCHANTMENT   
KARA_05_01hpheroic                 NEUTRAL   HERO_POWER   
BRMA10_3                           NEUTRAL   HERO_POWER   
TB_PickYourFate_10_EnchMinion      NEUTRAL  ENCHANTMENT   
NAX3_02H                           NEUTRAL   HERO_POWER   
AT_083e                            NEUTRAL  ENCHANTMENT   
AT_066e                            WARRIOR  ENCHANTMENT   
CS2_031                               MAGE        SPELL   
OG_195e                            NEUTRAL  ENCHANTMENT   
XXX_111e                           NEUTRAL  ENCHANTMENT   
NAX3_01                            NEUTRAL         HERO   

                                                                 name  \
card_id                                                                 
KARA_00_07                                              Astral Portal   
NEW1_008a                                           Ancient Teachings   
BRM_010t2           

In [26]:
dust_df_indexed = dust_df.set_index('card_id', True)
dust_df_indexed

action  cost
card_id                           
BRM_010t2    CRAFTING_NORMAL    40
BRM_010t2    CRAFTING_GOLDEN   400
BRM_010t2  DISENCHANT_NORMAL     5
BRM_010t2  DISENCHANT_GOLDEN    50
AT_132       CRAFTING_NORMAL  1600
AT_132       CRAFTING_GOLDEN  3200
AT_132     DISENCHANT_NORMAL   400
AT_132     DISENCHANT_GOLDEN  1600
OG_141       CRAFTING_NORMAL    40
OG_141       CRAFTING_GOLDEN   400
OG_141     DISENCHANT_NORMAL     5
OG_141     DISENCHANT_GOLDEN    50
AT_131       CRAFTING_NORMAL  1600
AT_131       CRAFTING_GOLDEN  3200
AT_131     DISENCHANT_NORMAL   400
AT_131     DISENCHANT_GOLDEN  1600
EX1_583      CRAFTING_NORMAL    40
EX1_583      CRAFTING_GOLDEN   400
EX1_583    DISENCHANT_NORMAL     5
EX1_583    DISENCHANT_GOLDEN    50
CFM_095      CRAFTING_NORMAL   400
CFM_095      CRAFTING_GOLDEN  1600
CFM_095    DISENCHANT_NORMAL   100
CFM_095    DISENCHANT_GOLDEN   400
BRM_018      CRAFTING_NORMAL   100
BRM_018      CRAFTING_GOLDEN   800
BRM_018    DISENCHANT_NORMAL    20
BRM_018    DISENCHANT_GOLDEN   100
OG_027       CRAFTING_NORMAL   100
OG_027       CRAFTING_GOLDEN   800
...                      ...   ...
tt_010a    DISENCHANT_NORMAL   100
tt_010a    DISENCHANT_GOLDEN   400
EX1_559      CRAFTING_NORMAL  1600
EX1_559      CRAFTING_GOLDEN  3200
EX1_559    DISENCHANT_NORMAL   400
EX1_559    DISENCHANT_GOLDEN  1600
EX1_624      CRAFTING_NORMAL   100
EX1_624      CRAFTING_GOLDEN   800
EX1_624    DISENCHANT_NORMAL    20
EX1_624    DISENCHANT_GOLDEN   100
OG_044a      CRAFTING_NORMAL    40
OG_044a      CRAFTING_GOLDEN   400
OG_044a    DISENCHANT_NORMAL     5
OG_044a    DISENCHANT_GOLDEN    50
AT_044       CRAFTING_NORMAL   400
AT_044       CRAFTING_GOLDEN  1600
AT_044     DISENCHANT_NORMAL   100
AT_044     DISENCHANT_GOLDEN   400
EX1_345t     CRAFTING_NORMAL   400
EX1_345t     CRAFTING_GOLDEN  1600
EX1_345t   DISENCHANT_NORMAL   100
EX1_345t   DISENCHANT_GOLDEN   400
EX1_009      CRAFTING_NORMAL   100
EX1_009      CRAFTING_GOLDEN   800
EX1_009    DISENCHANT_NORMAL    20
EX1_009    DISENCHANT_GOLDEN   100
CS2_031      CRAFTING_NORMAL    40
CS2_031      CRAFTING_GOLDEN   400
CS2_031    DISENCHANT_NORMAL     5
CS2_031    DISENCHANT_GOLDEN    50

[4040 rows x 2 columns]

In [28]:
entourages_df_indexed = entourages_df.set_index('card_id', True)
entourages_df_indexed

entourage_card_id
card_id                                       
KAR_A10_22                          KAR_A10_09
KAR_A10_22                          KAR_A10_02
KAR_A10_22                          KAR_A10_08
KAR_A10_22                          KAR_A10_04
KAR_A10_22                          KAR_A10_05
TBST_006                              TBST_003
TBST_006                              TBST_002
TBST_006                              TBST_001
BRMA14_10H                            BRMA14_3
BRMA14_10H                           BRMA14_5H
BRMA14_10H                           BRMA14_7H
BRMA14_10H                           BRMA14_9H
BRMA09_3                              BRMA09_2
BRMA09_3                              BRMA09_4
BRMA09_3                              BRMA09_5
KARA_06_03hp                        KARA_06_01
BRMA12_2                              BRMA12_6
BRMA12_2                              BRMA12_5
BRMA12_2                              BRMA12_7
BRMA12_2                              BRMA12_4
BRMA12_2                              BRMA12_3
TB_PickYourFate_2nd          TB_PickYourFate_1
TB_PickYourFate_2nd          TB_PickYourFate_4
TB_PickYourFate_2nd      TB_PickYourFate_7_2nd
TB_PickYourFate_2nd          TB_PickYourFate_3
TB_PickYourFate_2nd      TB_PickYourFate_6_2nd
TB_PickYourFate_2nd     TB_PickYourFate_11rand
TB_001                               BRMA14_10
TB_001                              BRMA14_10H
TB_001                                BRMA09_4
...                                        ...
LOEA16_16H                           LOEA16_10
LOEA16_16H                           LOEA16_11
LOEA16_16H                           LOEA16_14
LOEA16_16H                           LOEA16_15
LOEA16_16H                            LOEA16_6
LOEA16_16H                            LOEA16_7
LOEA16_16H                            LOEA16_9
LOEA16_16H                           LOEA16_12
LOEA16_16H                           LOEA16_13
LOEA16_16H                            LOEA16_8
TBUD_1                                 CS1_042
TBUD_1                                 CS2_171
BRMA13_3H                             BRMA13_8
BRMA13_3H                             BRMA13_7
BRMA13_3H                             BRMA13_6
BRMA13_3H                             BRMA13_5
TB_MechWar_CommonCards                 GVG_069
TB_MechWar_CommonCards                 GVG_079
TB_MechWar_CommonCards                 GVG_034
TB_MechWar_CommonCards                 GVG_048
TB_MechWar_CommonCards                 GVG_049
TB_MechWar_CommonCards                 GVG_088
KARA_00_10                             EX1_294
KARA_00_10                             FP1_018
KARA_00_10                              AT_002
KARA_00_10                             EX1_287
KARA_00_10                              tt_010
KARA_00_10                             EX1_295
KARA_00_10                             EX1_289
KARA_00_10                             EX1_594

[328 rows x 1 columns]

In [30]:
mechanics_df_indexed = mechanics_df.set_index('card_id', True)
mechanics_df_indexed

mechanic
card_id                                
AT_132                        BATTLECRY
GVG_011a            TAG_ONE_TURN_EFFECT
EX1_583                       BATTLECRY
LOE_007t                      EVIL_GLOW
LOE_007t             ImmuneToSpellpower
CFM_095                     DEATHRATTLE
BRM_018                       BATTLECRY
GVG_030                      CHOOSE_ONE
GVG_030                           TAUNT
CS2_227                            AURA
KAR_062                       BATTLECRY
CFM_712_t17                  JADE_GOLEM
BRM_011t            TAG_ONE_TURN_EFFECT
BRMA10_3H                  AI_MUST_PLAY
KAR_710m                          TAUNT
EX1_295o            TAG_ONE_TURN_EFFECT
OG_158                      DEATHRATTLE
LOE_110t                        TOPDECK
LOE_110t             ImmuneToSpellpower
BRMA13_4H                  AI_MUST_PLAY
EX1_145o            TAG_ONE_TURN_EFFECT
LOEA09_2e           TAG_ONE_TURN_EFFECT
EX1_570e            TAG_ONE_TURN_EFFECT
GVG_069                       BATTLECRY
hexfrog                           TAUNT
OG_145                    DIVINE_SHIELD
OG_145                            TAUNT
AT_090                          INSPIRE
AT_060                           SECRET
TB_MechWar_Minion1                TAUNT
...                                 ...
EX1_284                       BATTLECRY
LOE_038                            AURA
AT_083                          INSPIRE
TB_CoOpv3_BOSS2e                ENRAGED
GVG_009                       BATTLECRY
CFM_338                       BATTLECRY
GVG_097                       BATTLECRY
GVG_097                           TAUNT
CFM_852                       BATTLECRY
LOE_020                       BATTLECRY
CFM_781                         STEALTH
GVG_090                       BATTLECRY
GVG_120                       BATTLECRY
EX1_055o            TAG_ONE_TURN_EFFECT
CFM_321                       BATTLECRY
EX1_082                       BATTLECRY
FP1_017                            AURA
FP1_001                     DEATHRATTLE
AT_089                          INSPIRE
CFM_646                     DEATHRATTLE
AT_084                        BATTLECRY
OG_044a                          CHARGE
OG_044a                           TAUNT
LOEA09_3d                  AI_MUST_PLAY
BRMA13_4                   AI_MUST_PLAY
EX1_009                         ENRAGED
BRMA10_3                   AI_MUST_PLAY
NAX3_02H                   AI_MUST_PLAY
AT_083e             TAG_ONE_TURN_EFFECT
CS2_031                          FREEZE

[1079 rows x 1 columns]

In [31]:
play_requirements_df_indexed = play_requirements_df.set_index('card_id', True)
play_requirements_df_indexed

play_requirement  value
card_id                                                
KARA_00_07                  REQ_NUM_MINION_SLOTS      1
PRO_001a                    REQ_NUM_MINION_SLOTS      1
NAX1_01                     REQ_NUM_MINION_SLOTS      1
DS1h_292_H1                      REQ_STEADY_SHOT      0
DS1h_292_H1             REQ_MINION_OR_ENEMY_HERO      0
KAR_076                       REQ_TARGET_TO_PLAY      0
AT_024                        REQ_TARGET_TO_PLAY      0
AT_024                      REQ_TARGET_WITH_RACE     15
AT_024                         REQ_MINION_TARGET      0
BRMA06_2                    REQ_NUM_MINION_SLOTS      1
NAX5_02                REQ_MINIMUM_ENEMY_MINIONS      1
OG_080c                      REQ_FRIENDLY_TARGET      0
OG_080c                       REQ_TARGET_TO_PLAY      0
OG_080c                        REQ_MINION_TARGET      0
CS2_039                       REQ_TARGET_TO_PLAY      0
CS2_039                        REQ_MINION_TARGET      0
TB_KaraPortal_002              REQ_MINION_TARGET      0
CS2_094                       REQ_TARGET_TO_PLAY      0
TB_SPT_DPromoHP                REQ_MINION_TARGET      0
BRMA15_3                    REQ_NUM_MINION_SLOTS      1
CS2_141                  REQ_TARGET_IF_AVAILABLE      0
CS2_141                       REQ_NONSELF_TARGET      0
CS2_063                       REQ_TARGET_TO_PLAY      0
CS2_063                        REQ_MINION_TARGET      0
CS2_063                         REQ_ENEMY_TARGET      0
GVG_055e                     REQ_FRIENDLY_TARGET      0
TB_006                        REQ_TARGET_TO_PLAY      0
TB_006                         REQ_MINION_TARGET      0
CFM_621t21             REQ_MINIMUM_TOTAL_MINIONS      1
LOEA06_04h                    REQ_TARGET_TO_PLAY      0
...                                          ...    ...
GVG_011                        REQ_MINION_TARGET      0
CFM_061                       REQ_TARGET_TO_PLAY      0
AT_132_HUNTER                    REQ_STEADY_SHOT      0
AT_132_HUNTER           REQ_MINION_OR_ENEMY_HERO      0
GVG_047                        REQ_MINION_TARGET      0
GVG_047                         REQ_ENEMY_TARGET      0
EX1_278                       REQ_TARGET_TO_PLAY      0
CS2_012                       REQ_TARGET_TO_PLAY      0
CS2_012                         REQ_ENEMY_TARGET      0
EX1_578                       REQ_TARGET_TO_PLAY      0
EX1_578                        REQ_MINION_TARGET      0
KARA_00_10         REQ_SECRET_CAP_FOR_NON_SECRET      0
KARA_13_11             REQ_MINIMUM_ENEMY_MINIONS      2
GVG_120                  REQ_TARGET_IF_AVAILABLE      0
GVG_120                     REQ_TARGET_WITH_RACE     20
LOEA02_10a                  REQ_NUM_MINION_SLOTS      1
CS1_130                       REQ_TARGET_TO_PLAY      0
AT_025                 REQ_MINIMUM_ENEMY_MINIONS      2
CS2_061                       REQ_TARGET_TO_PLAY      0
AT_084                       REQ_FRIENDLY_TARGET      0
AT_084                   REQ_TARGET_IF_AVAILABLE      0
AT_084                         REQ_MINION_TARGET      0
BRMA16_3                     REQ_WEAPON_EQUIPPED      0
BRMA16_3                      REQ_TARGET_TO_PLAY      0
BRMA16_3                       REQ_MINION_TARGET      0
EX1_624                       REQ_TARGET_TO_PLAY      0
AT_044                        REQ_TARGET_TO_PLAY      0
AT_044                         REQ_MINION_TARGET      0
AT_132_SHAMAN               REQ_NUM_MINION_SLOTS      1
CS2_031                       REQ_TARGET_TO_PLAY      0

[776 rows x 2 columns]

### Executing Joins

Now that we have the tables loaded correctly, we're going to execute some joins. There are 4 different kinds of joins, which can best be visualized with venn diagrams:

<img src='joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table.  The left table is the table we are working with, and the right table is the table we want to join to the table we are working with.  We'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table we want to join with, `mechanics_df`
* The `how` parameter set to the type of join we want, `'inner'`

In [41]:
to_concat2 = [cards_df_indexed, mechanics_df_indexed]

cards_with_mechanics_df = pd.concat(to_concat2, axis=1, join = 'inner')
cards_with_mechanics_df

player_class         type                   name      set  \
card_id                                                                        
AT_132                  NEUTRAL       MINION     Justicar Trueheart      TGT   
GVG_011a                NEUTRAL  ENCHANTMENT             Shrink Ray      GVG   
EX1_583                 NEUTRAL       MINION     Priestess of Elune  EXPERT1   
LOE_007t                WARLOCK        SPELL                Cursed!      LOE   
LOE_007t                WARLOCK        SPELL                Cursed!      LOE   
CFM_095                 NEUTRAL       MINION        Weasel Tunneler    GANGS   
BRM_018                 PALADIN       MINION         Dragon Consort      BRM   
GVG_030                   DRUID       MINION      Anodized Robo Cub      GVG   
GVG_030                   DRUID       MINION      Anodized Robo Cub      GVG   
CS2_227                 NEUTRAL       MINION  Venture Co. Mercenary  EXPERT1   
KAR_062                 NEUTRAL       MINION  Netherspite Historian     KARA   
CFM_712_t17             NEUTRAL       MINION             Jade Golem    GANGS   
BRM_011t                 SHAMAN  ENCHANTMENT             Lava Shock      BRM   
BRMA10_3H               NEUTRAL   HERO_POWER            The Rookery      BRM   
KAR_710m                NEUTRAL       MINION        Animated Shield     KARA   
EX1_295o                   MAGE  ENCHANTMENT              Ice Block  EXPERT1   
OG_158                  NEUTRAL       MINION       Zealous Initiate       OG   
LOE_110t                NEUTRAL        SPELL          Ancient Curse      LOE   
LOE_110t                NEUTRAL        SPELL          Ancient Curse      LOE   
BRMA13_4H               NEUTRAL   HERO_POWER             Wild Magic      BRM   
EX1_145o                  ROGUE  ENCHANTMENT            Preparation  EXPERT1   
LOEA09_2e               NEUTRAL  ENCHANTMENT                Enraged      LOE   
EX1_570e                NEUTRAL  ENCHANTMENT                   Bite  EXPERT1   
GVG_069                 NEUTRAL       MINION        Antique Healbot      GVG   
hexfrog                 NEUTRAL       MINION                   Frog     CORE   
OG_145                  NEUTRAL       MINION           Psych-o-Tron       OG   
OG_145                  NEUTRAL       MINION           Psych-o-Tron       OG   
AT_090                  NEUTRAL       MINION       Mukla's Champion      TGT   
AT_060                   HUNTER        SPELL              Bear Trap      TGT   
TB_MechWar_Minion1      NEUTRAL       MINION               Mech Fan       TB   
...                         ...          ...                    ...      ...   
EX1_284                 NEUTRAL       MINION            Azure Drake  EXPERT1   
LOE_038                 NEUTRAL       MINION         Naga Sea Witch      LOE   
AT_083                  NEUTRAL       MINION       Dragonhawk Rider      TGT   
TB_CoOpv3_BOSS2e        NEUTRAL  ENCHANTMENT      Getting Angry....       TB   
GVG_009                  PRIEST       MINION           Shadowbomber      GVG   
CFM_338                  HUNTER       MINION       Trogg Beastrager    GANGS   
GVG_097                 NEUTRAL       MINION          Lil' Exorcist      GVG   
GVG_097                 NEUTRAL       MINION          Lil' Exorcist      GVG   
CFM_852                 NEUTRAL       MINION           Lotus Agents    GANGS   
LOE_020                  HUNTER       MINION           Desert Camel      LOE   
CFM_781                   ROGUE       MINION   Shaku, the Collector    GANGS   
GVG_090                 NEUTRAL       MINION          Madder Bomber      GVG   
GVG_120                 NEUTRAL       MINION       Hemet Nesingwary      GVG   
EX1_055o                NEUTRAL  ENCHANTMENT              Empowered  EXPERT1   
CFM_321                 NEUTRAL       MINION  Grimestreet Informant    GANGS   
EX1_082                 NEUTRAL       MINION             Mad Bomber  EXPERT1   
FP1_017                 NEUTRAL       MINION       Nerub'ar Weblord     NAXX   
FP1_001            

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads we displayed earlier.  Notice how it this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, conntains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________
'cards_with_mechanics_df' contains only those rows found it both cards_df AND mechanics_df

### Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [42]:
left_join_df = cards_with_mechanics_df.join(play_requirements_df_indexed)
left_join_df

player_class         type                        name  \
card_id                                                                      
AT_002                       MAGE        SPELL                      Effigy   
AT_005t                   NEUTRAL       MINION                        Boar   
AT_006                       MAGE       MINION            Dalaran Aspirant   
AT_007                       MAGE       MINION                Spellslinger   
AT_009                       MAGE       MINION                      Rhonin   
AT_010                     HUNTER       MINION                Ram Wrangler   
AT_011e                   NEUTRAL  ENCHANTMENT            Light's Blessing   
AT_012                     PRIEST       MINION            Spawn of Shadows   
AT_017                    NEUTRAL       MINION           Twilight Guardian   
AT_018                     PRIEST       MINION         Confessor Paletress   
AT_019                    WARLOCK       MINION                  Dreadsteed   
AT_022                    WARLOCK        SPELL            Fist of Jaraxxus   
AT_023                    WARLOCK       MINION                Void Crusher   
AT_028                      ROGUE       MINION             Shado-Pan Rider   
AT_030                      ROGUE       MINION           Undercity Valiant   
AT_032                      ROGUE       MINION                Shady Dealer   
AT_035t                     ROGUE        SPELL                     Ambush!   
AT_036                      ROGUE       MINION                   Anub'arak   
AT_037                      DRUID        SPELL                Living Roots   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_039                      DRUID       MINION            Savage Combatant   
AT_039e                     DRUID  ENCHANTMENT                      Savage   
AT_040                      DRUID       MINION                  Wildwalker   
AT_040                      DRUID       MINION                  Wildwalker   
AT_040                      DRUID       MINION                  Wildwalker   
AT_040                      DRUID       MINION                  Wildwalker   
AT_042                      DRUID       MINION          Druid of the Saber   
AT_042t                     DRUID       MINION             Sabertooth Lion   
AT_042t2                    DRUID       MINION          Sabertooth Panther   
...                           ...          ...                         ...   
TB_SPT_DPromoSecret3      WARRIOR        SPELL            Visions of Valor   
TB_SPT_DPromoSecret4      WARRIOR        SPELL             Visions of Fate   
TB_SPT_DPromoSecret5      WARRIOR        SPELL       Visions of the Amazon   
TB_SPT_DPromoSecret6      WARRIOR        SPELL     Visions of the Sorcerer   
TB_SPT_DPromoSecret7      WARRIOR        SPELL  Visions of the Necromancer   
TB_SPT_DPromoSecret9      WARRIOR        SPELL        Visions of Knowledge   
TB_SPT_DPromo_Hero2       WARRIOR         HERO                The Cow King   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_Minion1            NEUTRAL       MINION                  Shieldsman   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TU4c_003                  NEUTRAL       MINION                      Barrel   
TU4c_008e                 NEUTRAL  ENCHANTMENT              Might of Mukla   
TU4f_007                  NEUTRAL       MINION                Crazy Monkey   
XXX_001                   NEUTRAL        SPELL                    Damage 1   
XXX_002                   NEUTRAL        SPELL                    Damage 5   
XXX_008                   NEUTRAL        SPELL                      Freeze   
XXX_020                   NEUTRAL        SP

Note that the results of this sort of join are dependent upon the position of each table--if we were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________
play_requirements_df: play requirement, value columns, cards_with_mechanics_df: all other columns


#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [43]:
outer_join_df = cards_df.join(dust_df,rsuffix='_dust',how='outer')
outer_join_df

card_id player_class         type  \
0                 KARA_00_07         MAGE        SPELL   
1                  NEW1_008a        DRUID        SPELL   
2                  BRM_010t2        DRUID       MINION   
3                     AT_132      NEUTRAL       MINION   
4                     OG_141      NEUTRAL       MINION   
5                    CRED_17      NEUTRAL       MINION   
6                    OG_337e      NEUTRAL  ENCHANTMENT   
7                  BRMA08_1H      NEUTRAL         HERO   
8                   GVG_011a      NEUTRAL  ENCHANTMENT   
9                NAX8_02H_TB      NEUTRAL   HERO_POWER   
10                    AT_131      NEUTRAL       MINION   
11                  CS2_122e      NEUTRAL  ENCHANTMENT   
12               LOEA05_02ha      NEUTRAL   HERO_POWER   
13         CS2_102_H1_AT_132      WARRIOR   HERO_POWER   
14                  CFM_334e      NEUTRAL  ENCHANTMENT   
15                   EX1_583      NEUTRAL       MINION   
16    TB_PickYourFate_11rand      NEUTRAL        SPELL   
17                  LOE_007t      WARLOCK        SPELL   
18                   CFM_095      NEUTRAL       MINION   
19                   BRM_018      PALADIN       MINION   
20                  PRO_001a      NEUTRAL        SPELL   
21                 BRMA10_5H      NEUTRAL       MINION   
22                  BRMA11_1      NEUTRAL         HERO   
23                    OG_027       SHAMAN        SPELL   
24                   NAX1_01      NEUTRAL         HERO   
25                   GVG_030        DRUID       MINION   
26                   CFM_094      WARLOCK        SPELL   
27               DS1h_292_H1       HUNTER   HERO_POWER   
28                   CS2_227      NEUTRAL       MINION   
29                   KAR_062      NEUTRAL       MINION   
...                      ...          ...          ...   
4010                     NaN          NaN          NaN   
4011                     NaN          NaN          NaN   
4012                     NaN          NaN          NaN   
4013                     NaN          NaN          NaN   
4014                     NaN          NaN          NaN   
4015                     NaN          NaN          NaN   
4016                     NaN          NaN          NaN   
4017                     NaN          NaN          NaN   
4018                     NaN          NaN          NaN   
4019                     NaN          NaN          NaN   
4020                     NaN          NaN          NaN   
4021                     NaN          NaN          NaN   
4022                     NaN          NaN          NaN   
4023                     NaN          NaN          NaN   
4024                     NaN          NaN          NaN   
4025                     NaN          NaN          NaN   
4026                     NaN          NaN          NaN   
4027                     NaN          NaN          NaN   
4028                     NaN          NaN          NaN   
4029                     NaN          NaN          NaN   
4030                     NaN          NaN          NaN   
4031                     NaN          NaN          NaN   
4032                     NaN          NaN          NaN   
4033                     NaN          NaN          NaN   
4034                     NaN          NaN          NaN   
4035                     NaN          NaN          NaN   
4036                     NaN          NaN          NaN   
4037                     NaN          NaN          NaN   
4038                     NaN          NaN          NaN   
4039                     NaN          NaN          NaN   

                         name         set  \
0               Astral Portal        KARA   
1           Ancient Teachings     EXPERT1   
2          Druid of the Flame         BRM   
3          Justicar Trueheart         TGT   
4           Faceless Behemoth          OG   
5                   Rob Pardo     CREDITS   
6          Eve of Destruction          OG   
7          General Drakkisath         BRM   
8                  

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, we learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Use `sqlalchemy` to connect to a sqlite database and read in individual tables as DataFrames
* Understand and execute the various types of joins (inner, outer, left, and right joins)